log4j-api/src/main/java/org/apache/logging/log4j/util/OsgiServiceLocator.java

```diff
@@ -57,29 +58,53 @@ public static <T> Stream<T> loadServices(final Class<T> serviceType, final Looku
 
     public static <T> Stream<T> loadServices(final Class<T> serviceType, final Lookup lookup, final boolean verbose) {
         final Class<?> lookupClass = Objects.requireNonNull(lookup, "lookup").lookupClass();
-        final Bundle bundle = FrameworkUtil.getBundle(Objects.requireNonNull(lookupClass, "lookupClass"));
-        if (bundle != null) {
+        return loadServices(serviceType, lookupClass, StatusLogger.getLogger());
+    }
+
+    static <T> Stream<T> loadServices(final Class<T> serviceType, final Class<?> callerClass, final Logger logger) {
+        final Bundle bundle = FrameworkUtil.getBundle(callerClass);
+        if (bundle != null && !isFragment(bundle)) {
             final BundleContext ctx = bundle.getBundleContext();
             if (ctx == null) {
-                if (verbose) {
-                    StatusLogger.getLogger()
-                            .error(
-                                    "Unable to load OSGI services: The bundle has no valid BundleContext for serviceType = {}, lookup = {}, lookupClass = {}, bundle = {}",
-                                    serviceType,
-                                    lookup,
-                                    lookupClass,
-                                    bundle);
-                }
+                logger.warn(
+                        "Unable to load OSGi services for service {}: bundle {} (state {}) does not have a valid BundleContext",
+                        serviceType::getName,
+                        bundle::getSymbolicName,
+                        () -> {
+                            switch (bundle.getState()) {
+                                case Bundle.UNINSTALLED:
+                                    return "UNINSTALLED";
+                                case Bundle.INSTALLED:
+                                    return "INSTALLED";
+                                case Bundle.RESOLVED:
+                                    return "RESOLVED";
+                                case Bundle.STARTING:
+                                    return "STARTING";
+                                case Bundle.STOPPING:
+                                    return "STOPPING";
+                                case Bundle.ACTIVE:
+                                    return "ACTIVE";
+                                default:
+                                    return "UNKNOWN";
+                            }
+                        });
```

Comment:
I can't stop but wonder, why not simply use `bundle.getState()` and assume its `toString()` will do the right job.

Reply:
`bundle.getState()` is an integer. ;-)

Here I wanted to provide some debugging data for us in case `OsgiServiceLocator` starts misbehaving again, due to the changes in this PR.